arwen/contexts/async.go

```diff
@@ -0,0 +1,885 @@
+func (context *asyncContext) HasPendingCallGroups() bool {
+	return len(context.asyncCallGroups) > 0
+}
+
+// IsComplete returns true if there are no more AsyncCallGroups contained in the AsyncContext.
+func (context *asyncContext) IsComplete() bool {
+	return len(context.asyncCallGroups) == 0
+}
+
+// Save serializes and saves the AsyncContext to the storage of the contract, under a protected key.
+func (context *asyncContext) Save() error {
+	if len(context.asyncCallGroups) == 0 {
+		return nil
+	}
+
+	storage := context.host.Storage()
+	runtime := context.host.Runtime()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetPrevTxHash())
+	data, err := context.Serialize()
+	if err != nil {
+		return err
+	}
+
+	_, err = storage.SetProtectedStorage(storageKey, data)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Load restores the internal state of the AsyncContext from the storage of the contract.
+func (context *asyncContext) Load() error {
+	runtime := context.host.Runtime()
+	storage := context.host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetPrevTxHash())
+	data := storage.GetStorage(storageKey)
+	if len(data) == 0 {
+		return arwen.ErrNoStoredAsyncContextFound
+	}
+
+	loadedContext, err := context.deserialize(data)
+	if err != nil {
+		return err
+	}
+
+	context.callerAddr = loadedContext.callerAddr
+	context.returnData = loadedContext.returnData
+	context.asyncCallGroups = loadedContext.asyncCallGroups
+
+	return nil
+}
+
+// Delete deletes the persisted state of the AsyncContext from the contract storage.
+func (context *asyncContext) Delete() error {
+	runtime := context.host.Runtime()
+	storage := context.host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetPrevTxHash())
+	_, err := storage.SetProtectedStorage(storageKey, nil)
+	return err
+}
+
+func (context *asyncContext) determineExecutionMode(destination []byte, data []byte) (arwen.AsyncCallExecutionMode, error) {
+	runtime := context.host.Runtime()
+	blockchain := context.host.Blockchain()
+
+	// If ArgParser cannot read the Data field, then this is neither a SC call,
+	// nor a built-in function call.
+	functionName, args, err := context.host.CallArgsParser().ParseData(string(data))
+	if err != nil {
+		return arwen.AsyncUnknown, err
+	}
+
+	sameShard := context.host.AreInSameShard(runtime.GetSCAddress(), destination)
+	if context.host.IsBuiltinFunctionName(functionName) {
+		if sameShard {
+			vmInput := runtime.GetVMInput()
+			isESDTTransfer, _, _ := context.isESDTTransferOnReturnDataFromFunctionAndArgs(runtime.GetSCAddress(), destination, functionName, args)
+			isAsyncCall := vmInput.CallType == vm.AsynchronousCall
+			isReturningCall := bytes.Equal(vmInput.CallerAddr, destination)
+
+			if isESDTTransfer && isAsyncCall && isReturningCall {
+				return arwen.ESDTTransferOnCallBack, nil
```

Comment:
1. Yes, it's so that contracts can send ESDT to their callers via async.
2. Very good question. Will investigate. The current implementation is designed to handle a single ESDT async callback. This will very likely need to be reworked.